packages/near-membrane-base/src/membrane.ts

```diff
@@ -754,17 +852,57 @@ export function createMembraneMarshall() {
                 this.preventExtensions = BoundaryProxyHandler.defaultPreventExtensionsTrap;
                 this.setPrototypeOf = BoundaryProxyHandler.defaultSetPrototypeOfTrap;
                 this.set = BoundaryProxyHandler.defaultSetTrap;
-                if (foreignTargetTraits & TargetTraits.Revoked) {
+                if (isRevoked) {
                     revoke();
+                } else if (isUnrevokedObject) {
+                    // Lazily define unbrandedTag.
+                    let unbrandedTag: string | undefined | symbol =
+                        LOCKER_NEAR_MEMBRANE_UNDEFINED_VALUE_SYMBOL;
+                    ReflectApply(ObjectProto__defineGetter__, this, [
+                        'unbrandedTag',
+                        () => {
+                            if (unbrandedTag === LOCKER_NEAR_MEMBRANE_UNDEFINED_VALUE_SYMBOL) {
+                                const toStringTag = foreignCallableGetToStringTagOfTarget(
+                                    this.foreignTargetPointer
+                                );
+                                // The default language toStringTag is "Object".
+                                // If receive "Object" we return `undefined` to
+                                // let the language resolve it naturally without
+                                // projecting a value.
+                                unbrandedTag = toStringTag === 'Object' ? undefined : toStringTag;
+                            }
+                            return unbrandedTag;
+                        },
+                    ]);
                 }
-                if (!trapMutations) {
+                if (trapMutations) {
+                    if (foreignTargetTraits & TargetTraits.IsArray) {
+                        this.makeProxyLive();
+                    }
+                } else {
+                    if (isUnrevokedObject) {
+                        // Lazily define serializedValue.
+                        let serializedValue: SerializedValue | undefined | symbol =
+                            LOCKER_NEAR_MEMBRANE_UNDEFINED_VALUE_SYMBOL;
+                        ReflectApply(ObjectProto__defineGetter__, this, [
```

Comment:
so, by adding new stuff to the handler... it seems that this will make the code a lot more complicated. Also, why can't this be well defined at the time of the creation of the handler? Why does it have to be an accessor and just not a regular method of the handler?

Reply:
> so, by adding new stuff to the handler... it seems that this will make the code a lot more complicated.

It's 2 deferred properties. We could use a `state` or `data` property on the handler and bolt on to that to avoid issues with it being frozen.

> Also, why can't this be well defined at the time of the creation of the handler

That puts a tax on all relevant objects for data that is not needed in many cases.